HealthGPT/HealthGPT/HealthDataFetcher.swift

```diff
@@ -64,120 +67,151 @@ class HealthDataFetcher {
+                if let error = error {
+                    continuation.resume(throwing: error)
+                } else if let statsCollection = results {
+                    var dailyData: [Double] = []
+
+                    statsCollection.enumerateStatistics(
+                        from: Date().twoWeeksAgoStartOfDay(),
+                        to: Date.startOfDay()
+                    ) { statistics, _ in
+                        if let quantity = statistics.sumQuantity() {
+                            dailyData.append(quantity.doubleValue(for: unit))
+                        } else {
+                            dailyData.append(0)
+                        }
                     }
-                }
 
-                completion(dailyData)
-            } else {
-                completion([])
+                    continuation.resume(returning: dailyData)
+                } else {
+                    continuation.resume(throwing: HealthDataFetcherError.resultsNotFound)
+                }
             }
-        }
 
-        healthStore.execute(query)
+            healthStore.execute(query)
+        }
     }
 
+    /// Fetches the user's health data for the specified category type identifier for the last two weeks.
+    ///
+    /// - Parameter identifier: The `HKCategoryTypeIdentifier` representing the type of health data to fetch.
+    /// - Returns: An array of `Double` values representing the daily health data for the specified identifier.
+    /// - Throws: `HealthDataFetcherError` if the data cannot be fetched.
     func fetchLastTwoWeeksCategoryData(
-        for identifier: HKCategoryTypeIdentifier,
-        completion: @escaping ([Double]) -> Void
-    ) {
+        for identifier: HKCategoryTypeIdentifier
+    ) async throws -> [Double] {
         let predicate = createLastTwoWeeksPredicate()
 
         guard let sampleType = HKObjectType.categoryType(forIdentifier: identifier) else {
-            return
+            throw HealthDataFetcherError.invalidObjectType
         }
 
-        let query = HKSampleQuery(
-            sampleType: sampleType,
-            predicate: predicate,
-            limit: HKObjectQueryNoLimit,
-            sortDescriptors: nil
-        ) { _, samples, _ in
-            var dailyData: [Double] = [Double](repeating: 0, count: 14)
-            if let samples = samples as? [HKCategorySample] {
-                for sample in samples {
-                    let startOfSampleDay = Calendar.current.startOfDay(for: sample.startDate)
-                    let distance = Int(Date().timeIntervalSince(startOfSampleDay) / 86400)
-                    guard let minutes = Calendar.current.dateComponents(
-                        [.minute],
-                        from: sample.startDate,
-                        to: sample.endDate
-                    ).minute else {
-                        return
+        return try await withCheckedThrowingContinuation { continuation in
```

Comment:
Similar to the implementation above, using `HKSampleQueryDescriptor` https://developer.apple.com/documentation/healthkit/hksamplequerydescriptor might be a good point here.

Reply:
Thanks @PSchmiedmayer, I'm having some issues getting this function to work. If you could take another look for me, would really appreciate it.

Reply:
@vishnuravi I will take a look at this in more detail and will get back to you 👍